## Claude Review (Sonnet 独立セッション)

### 検証実施項目
- [x] PR diff 全件確認 (3ファイル: validate-samples.ts / processFlowAntipatternValidator.ts / .test.ts)
- [x] retail 落とし穴 16/17/19/23 の memory との整合確認
- [x] 13 ケース vitest 実測 pass 確認 (PASS 13 / FAIL 0 — 全体 1360 テスト PASS)
- [x] schemas/*.json 変更なし確認 (gh pr diff 743 --name-only に schemas/ は 0 件)
- [x] validate:samples ../samples/retail 実測確認 (**0 件ではない。後述 Must-fix 参照**)
- [x] 半角カナ文字化けチェック — `grep -P '[ｦ-ﾟ]'` 0 件 ✓

---

### Must-fix

**[M-1] retail サンプルが processFlowAntipatternValidator で 6 errors + 1 warning を出す**

PR 本文「retail サンプルは修正済のため 0 件」は誤りです。PRブランチ全依存ファイルをチェックアウトして `tsx validate-samples.ts ../samples/retail` を実測したところ、`processFlowAntipatternValidator` に起因する以下が出力されました:

```
[ERROR] [processFlowAntipatternValidator] [DUPLICATE_KIND_KEY]
  samples/retail/actions/d63c703c-...json (<raw offset 8990>)

[ERROR] [processFlowAntipatternValidator] [LITERAL_CONV_REFERENCE]
  samples/retail/actions/efa7ac6e-...json
  actions[0].steps[2].branches[0].steps[0].bodyExpression (2 件)

[ERROR] [processFlowAntipatternValidator] [INVALID_SEQUENCE_CALL_SYNTAX]
  samples/retail/actions/f81dd9e0-...json  actions[0].steps[4].expression

[ERROR] [processFlowAntipatternValidator] [LITERAL_CONV_REFERENCE]
  samples/retail/actions/f81dd9e0-...json (2 件)

[WARN] [processFlowAntipatternValidator] [MULTIPLE_STATEMENTS_IN_SQL]
  samples/retail/actions/f81dd9e0-...json  actions[0].steps[5].steps[3].sql
```

ISSUE #741 受入基準「retail サンプルで 0 件」が未達成です。validator が正しく機能している (4 Check の検出ロジックは動作している) にもかかわらず、retail サンプル自体がまだ落とし穴パターンを含んでいます。

**対応方針**: validator の誤動作ではなく、retail サンプルの修正漏れです。`samples/retail/actions/` 内の 3 ファイルを各 Check の正しい形に修正してください。

---

### Should-fix

**[S-1] Check 23: PL/pgSQL `DO $$ BEGIN ... END $$` ブロックと文字列リテラル内の `;` を誤検出する**

`hasMultipleStatements` の実装は `sql.replace(/;\s*$/, "").includes("; ")` のみで、以下を false positive として警告します:

- `DO $$ BEGIN INSERT INTO x VALUES (1); END $$` — ブロック内の `;`
- `WHERE name = 'a;b'` — 文字列リテラル内の `;`

ISSUE #741 本文には「PL/pgSQL の DO $$ BEGIN ... END $$ 等の例外考慮を文字列パターンで」と明記されていますが、実装では例外処理がありません。

severity = warning のため マージブロッカーではないですが、false positive が発生するとサンプル品質が正しく評価されません。

対処案:
1. `$$ ... $$` 区間の `;` を除外するパターンを追加する
2. または severity を `warning` のまま維持しつつ、メッセージに「PL/pgSQL ブロック内の `;` は誤検出の場合があります」と注釈を加える

**[S-2] PR 本文の自己申告 file:line が実際のファイルとズレている**

PR 本文の仕様逐条突合:
- `collectStringValues` の行番号: 申告 `162-179` → 実際 `198-211` (19 行ズレ)
- Check 16 の `LITERAL_CONV_RE` 定義: 申告 `45-54` → 実際 `48-52`

AGENTS.md「仕様逐条突合の file:line は実測値で埋める」に反します。M-1 修正後に PR 本文を更新してください。

---

### Nit

**[N-1] Check 17: findDuplicateKindObjects が O(n²) の走査コスト**

外側ループが `{` を見つけたあと内側 while で JSON 末端まで走査し、外側ループは `i++` で 1 文字だけ進む設計です。内側で走査済みの範囲を外側が再処理するため、深くネストした JSON では O(n²) になります。バリデーション用途で JSON サイズが小さければ実害はありませんが、将来大規模 JSON に適用する際は注意が必要です。

**[N-2] `AntipatternIssue.validator` の literal 型が他の validator の issue 型と非対称**

`sqlColumnIssue` 等の既存 validator は `validator` フィールドを持たず、`validate-samples.ts` の `issue()` ヘルパーが付与します。`AntipatternIssue` だけが `validator` フィールドを持ち、`issue()` の第一引数と二重になっています。動作上は問題ありませんが API 設計が不一致です。

---

### 総評

Check 16/17/19/23 の検出ロジック自体は正しく実装されており (vitest 13 件 pass、アルゴリズム検証 pass)、schema 変更なし、半角カナなしです。ただし **PR の核心的な受入基準「retail サンプル 0 件」が未達成 (M-1)** であり、validator が正しく落とし穴を検出してしまっている状態でマージすると CI が壊れます。M-1 の retail サンプル修正が完了するまでマージ不可です。

--- a/designer/src/schemas/processFlowAntipatternValidator.test.ts
+++ b/designer/src/schemas/processFlowAntipatternValidator.test.ts
@@ -228,4 +228,95 @@ describe("Check 23: MULTIPLE_STATEMENTS_IN_SQL", () => {
     const found = issues.filter((i) => i.code === "MULTIPLE_STATEMENTS_IN_SQL");
     expect(found).toHaveLength(0);
   });
+
+  it("negative: 文字列リテラル内の `;` は単一文扱い (false positive 防止)", () => {
+    const step = {
+      kind: "dbAccess",
+      id: "step-1",
+      sql: "SELECT * FROM users WHERE name = 'a;b' AND status = 'ok'",
+    };
+    const flow = makeFlow([step]);
+    const rawJson = JSON.stringify(flow, null, 2);
+    const issues = checkAntipatterns(flow, rawJson);
+    const found = issues.filter((i) => i.code === "MULTIPLE_STATEMENTS_IN_SQL");
+    expect(found).toHaveLength(0);
+  });
+
+  it("negative: SQL 標準のエスケープシングルクォート '' を含む文字列リテラル", () => {
+    const step = {
+      kind: "dbAccess",
+      id: "step-1",
+      sql: "SELECT * FROM users WHERE comment = 'it''s a; test'",
+    };
+    const flow = makeFlow([step]);
+    const rawJson = JSON.stringify(flow, null, 2);
+    const issues = checkAntipatterns(flow, rawJson);
+    const found = issues.filter((i) => i.code === "MULTIPLE_STATEMENTS_IN_SQL");
+    expect(found).toHaveLength(0);
+  });
+
+  it("negative: 行コメント `--` 内の `;` は単一文扱い", () => {
+    const step = {
+      kind: "dbAccess",
+      id: "step-1",
+      sql: "SELECT id FROM orders -- TODO: optimize; index hint\nWHERE customer_id = @customerId",
+    };
+    const flow = makeFlow([step]);
+    const rawJson = JSON.stringify(flow, null, 2);
+    const issues = checkAntipatterns(flow, rawJson);
+    const found = issues.filter((i) => i.code === "MULTIPLE_STATEMENTS_IN_SQL");
+    expect(found).toHaveLength(0);
+  });
+
+  it("negative: ブロックコメント `/* */` 内の `;` は単一文扱い", () => {
+    const step = {
+      kind: "dbAccess",
+      id: "step-1",
+      sql: "SELECT id /* multi-line; with; semicolons */ FROM orders WHERE id = @orderId",
+    };
+    const flow = makeFlow([step]);
+    const rawJson = JSON.stringify(flow, null, 2);
+    const issues = checkAntipatterns(flow, rawJson);
+    const found = issues.filter((i) => i.code === "MULTIPLE_STATEMENTS_IN_SQL");
+    expect(found).toHaveLength(0);
+  });
+
+  it("negative: PL/pgSQL ドル引用 `$$ ... $$` 内の `;` は単一文扱い", () => {
+    const step = {
+      kind: "dbAccess",
+      id: "step-1",
+      sql: "DO $$ BEGIN UPDATE accounts SET balance = balance + 100; UPDATE accounts SET locked = true; END $$",
+    };
+    const flow = makeFlow([step]);
+    const rawJson = JSON.stringify(flow, null, 2);
+    const issues = checkAntipatterns(flow, rawJson);
+    const found = issues.filter((i) => i.code === "MULTIPLE_STATEMENTS_IN_SQL");
+    expect(found).toHaveLength(0);
+  });
+
+  it("negative: タグ付きドル引用 `$tag$ ... $tag$` 内の `;` は単一文扱い", () => {
+    const step = {
+      kind: "dbAccess",
+      id: "step-1",
+      sql: "CREATE FUNCTION f() RETURNS void AS $body$ BEGIN UPDATE x SET v = 1; END $body$ LANGUAGE plpgsql",
+    };
+    const flow = makeFlow([step]);
+    const rawJson = JSON.stringify(flow, null, 2);
+    const issues = checkAntipatterns(flow, rawJson);
+    const found = issues.filter((i) => i.code === "MULTIPLE_STATEMENTS_IN_SQL");
+    expect(found).toHaveLength(0);
+  });
+
+  it("positive: ドル引用ブロックの後に別文がある場合は複数文として検出", () => {
+    const step = {
+      kind: "dbAccess",
+      id: "step-1",
+      sql: "DO $$ BEGIN UPDATE x SET v = 1; END $$; SELECT 1",
+    };
+    const flow = makeFlow([step]);
+    const rawJson = JSON.stringify(flow, null, 2);
+    const issues = checkAntipatterns(flow, rawJson);
+    const found = issues.filter((i) => i.code === "MULTIPLE_STATEMENTS_IN_SQL");
+    expect(found.length).toBeGreaterThan(0);
+  });
 });

--- a/designer/src/schemas/processFlowAntipatternValidator.ts
+++ b/designer/src/schemas/processFlowAntipatternValidator.ts
@@ -152,10 +152,85 @@ function hasInvalidSequenceSyntax(value: string): boolean {
 
 /**
  * sql フィールド内にセミコロンで区切られた複数文が含まれるか検出する。
- * 末尾のセミコロンのみは許容 (= 末尾 ; を除去した後に ; が残れば複数文)。
+ *
+ * 文字列リテラル (`'...'`) / 行コメント (`-- ...`) / ブロックコメント (`/* ... *\/`) /
+ * PL/pgSQL ドル引用 (`$$ ... $$` / `$tag$ ... $tag$`) 内の `;` は false positive を
+ * 避けるため除外する。末尾のセミコロンのみも許容 (single statement の正常終端)。
  */
 function hasMultipleStatements(sql: string): boolean {
-  return sql.replace(/;\s*$/, "").includes(";");
+  // 末尾の空白 + セミコロンを剥がす
+  const trimmed = sql.replace(/[\s;]+$/, "");
+
+  let inString = false;
+  let inDollar = false;
+  let dollarTag = "";
+  let inLineComment = false;
+  let inBlockComment = false;
+
+  for (let i = 0; i < trimmed.length; i++) {
+    const ch = trimmed[i];
+    const next = trimmed[i + 1];
+
+    if (inLineComment) {
+      if (ch === "\n") inLineComment = false;
+      continue;
+    }
+    if (inBlockComment) {
+      if (ch === "*" && next === "/") {
+        inBlockComment = false;
+        i++;
+      }
+      continue;
+    }
+    if (inString) {
+      // SQL 標準のエスケープ '' (シングルクォート 2 連続)
+      if (ch === "'" && next === "'") {
+        i++;
+        continue;
+      }
+      if (ch === "'") inString = false;
+      continue;
+    }
+    if (inDollar) {
+      if (ch === "$" && trimmed.slice(i, i + dollarTag.length) === dollarTag) {
+        inDollar = false;
+        i += dollarTag.length - 1;
+        dollarTag = "";
+      }
+      continue;
+    }
+
+    // コメント / 文字列 / ドル引用の開始
+    if (ch === "-" && next === "-") {
+      inLineComment = true;
+      i++;
+      continue;
+    }
+    if (ch === "/" && next === "*") {
+      inBlockComment = true;
+      i++;
+      continue;
+    }
+    if (ch === "'") {
+      inString = true;
+      continue;
+    }
+    if (ch === "$") {
+      // $$ または $tag$ パターン
+      const m = trimmed.slice(i).match(/^\$([a-zA-Z_]\w*)?\$/);
+      if (m) {
+        dollarTag = m[0];
+        inDollar = true;
+        i += dollarTag.length - 1;
+        continue;
+      }
+    }
+
+    // クォート / コメント / ドル引用の外側で `;` 検出 → 複数文
+    if (ch === ";") return true;
+  }
+
+  return false;
 }
 
 // ─── walkSteps ──────────────────────────────────────────────────────────────

